File: v4/rust/iprange-livedb/src/publication/reservation_inspection.rs
//! Exact discovery of one restart-authoritative publication reservation.

use std::fs::File;

use crate::cancellation::CancellationToken;
use crate::contract::PAGE_SIZE;
use crate::error;
use crate::error::ErrorCode;
use crate::live_lock::{self, Mode};
use crate::mapping::Mapping;

use super::namespace::{Destination, Identity, Name, NamespaceError, Regular, ScanError};
use super::problem::Problem;
use super::reservation::{self, Header, Selected};
use super::result::AccessPolicy;
use super::{ArtifactKind, DirectoryRole};

const FILE_SIZE: u64 = (2 * PAGE_SIZE) as u64;
const OPERATION_LOCK: u64 = 0;
const PREFIX: &[u8] = b".iprange-reservation-";
const SUFFIX: &[u8] = b".tmp";
const ENCODED_ATTEMPT_LEN: usize = 32;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(super) enum Location {
    Private,
    Canonical,
}

#[derive(Debug)]
pub(super) struct Inspected {
    pub(super) name: Name,
    pub(super) file: File,
    pub(super) mapping: Mapping,
    pub(super) identity: Identity,
    pub(super) header: Header,
    pub(super) location: Location,
    pub(super) access: AccessPolicy,
}

impl Inspected {
    pub(super) fn verify(&self, destination: &Destination) -> Result<(), Problem> {
        require_available(
            destination,
            self.location,
            &self.name,
            self.identity,
            self.header,
        )?;
        destination
            .directory()
            .verify()
            .map_err(|error| Problem::namespace(&error))?;
        let name = match self.location {
            Location::Private => &self.name,
            Location::Canonical => destination.coordination(),
        };
        destination
            .directory()
            .verify_name(name, self.identity)
            .map_err(|error| Problem::namespace(&error))?;
        let selected = read_selected(&self.mapping).map_err(strict_record)?;
        if selected.header != self.header {
            return Err(conflict("publication reservation changed after inspection"));
        }
        Ok(())
    }

    pub(super) fn unlock_operation(&self) -> Result<(), Problem> {
        live_lock::unlock(&self.file, OPERATION_LOCK).map_err(|error| Problem::sdk(&error))
    }

    pub(super) fn relock_operation(
        &self,
        destination: &Destination,
        cancellation: &CancellationToken,
    ) -> Result<(), Problem> {
        require_available(
            destination,
            self.location,
            &self.name,
            self.identity,
            self.header,
        )?;
        lock_operation_file(&self.file, cancellation)?;
        self.verify(destination)
    }
}

pub(super) fn discover(
    destination: &Destination,
    cancellation: &CancellationToken,
) -> Result<Option<Inspected>, Problem> {
    if let Some(inspected) = canonical(destination, cancellation)? {
        return Ok(Some(inspected));
    }
    let found = scan_private(destination, cancellation)?;
    destination
        .directory()
        .require_absent(destination.coordination())
        .map_err(|_| conflict("coordination changed during reservation scan"))?;
    Ok(found)
}

pub(super) fn canonical(
    destination: &Destination,
    cancellation: &CancellationToken,
) -> Result<Option<Inspected>, Problem> {
    cancellation.check().map_err(|error| Problem::sdk(&error))?;
    #[cfg(target_os = "freebsd")]
    let Some(regular) = destination
        .directory()
        .open_regular_any_link(destination.coordination(), true)
        .map_err(|error| Problem::namespace(&error))?
    else {
        return Ok(None);
    };
    #[cfg(not(target_os = "freebsd"))]
    let Some(regular) = destination
        .directory()
        .open_regular(destination.coordination(), true)
        .map_err(|error| Problem::namespace(&error))?
    else {
        return Ok(None);
    };
    inspect_canonical(destination, regular, cancellation).map(Some)
}

pub(super) fn exact_private(
    destination: &Destination,
    expected: Header,
    cancellation: &CancellationToken,
) -> Result<Option<Inspected>, Problem> {
    cancellation.check().map_err(|error| Problem::sdk(&error))?;
    let name = destination
        .reservation_name(expected.attempt_id)
        .map_err(|error| Problem::namespace(&error))?;
    let Some(regular) = destination
        .directory()
        .open_regular(&name, true)
        .map_err(|error| Problem::namespace(&error))?
    else {
        return Ok(None);
    };
    require_available(
        destination,
        Location::Private,
        &name,
        regular.identity,
        expected,
    )?;
    lock_operation(&regular, cancellation)?;
    destination
        .directory()
        .verify_name(&name, regular.identity)
        .map_err(|error| Problem::namespace(&error))?;
    let mapping = map_reservation(&regular.file).map_err(strict_record)?;
    let selected = read_selected(&mapping).map_err(strict_record)?;
    require_bound(
        destination,
        selected.header,
        regular.identity,
        Some(expected.attempt_id),
    )?;
    if selected.header != expected {
        return Err(conflict("caller result and private reservation disagree"));
    }
    Ok(Some(inspected(
        name,
        regular,
        mapping,
        selected,
        Location::Private,
    )))
}

fn inspect_canonical(
    destination: &Destination,
    regular: Regular,
    cancellation: &CancellationToken,
) -> Result<Inspected, Problem> {
    let mapping = map_reservation(&regular.file).map_err(strict_record)?;
    let selected = read_selected(&mapping).map_err(strict_record)?;
    require_available(
        destination,
        Location::Canonical,
        destination.coordination(),
        regular.identity,
        selected.header,
    )?;
    lock_operation(&regular, cancellation)?;
    if read_selected(&mapping).map_err(strict_record)? != selected {
        return Err(conflict(
            "publication reservation changed while acquiring its lock",
        ));
    }
    require_bound(destination, selected.header, regular.identity, None)?;
    let private_name = destination
        .reservation_name(selected.header.attempt_id)
        .map_err(|error| Problem::namespace(&error))?;
    #[cfg(target_os = "freebsd")]
    destination
        .directory()
        .finish_noreplace_transition(&private_name, destination.coordination(), regular.identity)
        .map_err(|error| Problem::namespace(&error))?;
    destination
        .directory()
        .verify_name(destination.coordination(), regular.identity)
        .map_err(|error| Problem::namespace(&error))?;
    if read_selected(&mapping).map_err(strict_record)? != selected {
        return Err(conflict(
            "publication reservation changed during inspection",
        ));
    }
    destination
        .directory()
        .verify()
        .map_err(|error| Problem::namespace(&error))?;
    Ok(inspected(
        private_name,
        regular,
        mapping,
        selected,
        Location::Canonical,
    ))
}

fn scan_private(
    destination: &Destination,
    cancellation: &CancellationToken,
) -> Result<Option<Inspected>, Problem> {
    let mut found = None;
    let scanned = destination.directory().scan(|bytes| {
        cancellation.check().map_err(|error| Problem::sdk(&error))?;
        let Some(attempt_id) = parse_private_name(bytes) else {
            return Ok(());
        };
        let name = Name::new(bytes).map_err(|error| Problem::namespace(&error))?;
        let Some(candidate) = inspect_private(destination, name, attempt_id, cancellation)? else {
            return Ok(());
        };
        if found.is_some() {
            return Err(conflict(
                "multiple bound private publication reservations exist",
            ));
        }
        found = Some(candidate);
        Ok(())
    });
    match scanned {
        Ok(()) => Ok(found),
        Err(ScanError::Namespace(error)) => Err(Problem::namespace(&error)),
        Err(ScanError::Visitor(problem)) => Err(problem),
    }
}

fn inspect_private(
    destination: &Destination,
    name: Name,
    attempt_id: [u8; 16],
    cancellation: &CancellationToken,
) -> Result<Option<Inspected>, Problem> {
    let regular = match destination.directory().open_regular(&name, true) {
        Ok(Some(regular)) => regular,
        Ok(None) => return Ok(None),
        Err(error) if invalid_private_entry(&error) => return Ok(None),
        Err(error) => return Err(Problem::namespace(&error)),
    };
    super::gc_barrier::require_source_available(
        destination.directory(),
        attempt_id,
        1,
        ArtifactKind::PrivateReservation,
        DirectoryRole::Destination,
        &name,
        regular.identity,
    )?;
    let mapping = match map_reservation(&regular.file) {
        Ok(mapping) => mapping,
        Err(ReadError::Invalid) => return Ok(None),
        Err(ReadError::Sdk(error)) => return Err(Problem::sdk(&error)),
    };
    let selected = match read_selected(&mapping) {
        Ok(selected) => selected,
        Err(ReadError::Invalid) => return Ok(None),
        Err(ReadError::Sdk(error)) => return Err(Problem::sdk(&error)),
    };
    if require_bound(
        destination,
        selected.header,
        regular.identity,
        Some(attempt_id),
    )
    .is_err()
    {
        return Ok(None);
    }
    lock_operation(&regular, cancellation)?;
    destination
        .directory()
        .verify_name(&name, regular.identity)
        .map_err(|_| conflict("private reservation changed during inspection"))?;
    let rechecked = read_selected(&mapping).map_err(strict_record)?;
    if rechecked != selected {
        return Err(conflict("private reservation changed during inspection"));
    }
    Ok(Some(inspected(
        name,
        regular,
        mapping,
        selected,
        Location::Private,
    )))
}

fn require_available(
    destination: &Destination,
    location: Location,
    private_name: &Name,
    identity: Identity,
    header: Header,
) -> Result<(), Problem> {
    let (name, kind) = match location {
        Location::Private => (private_name, ArtifactKind::PrivateReservation),
        Location::Canonical => (destination.coordination(), ArtifactKind::OwnedCoordination),
    };
    super::gc_barrier::require_source_available(
        destination.directory(),
        header.attempt_id,
        1,
        kind,
        DirectoryRole::Destination,
        name,
        identity,
    )
}

fn inspected(
    name: Name,
    regular: Regular,
    mapping: Mapping,
    selected: Selected,
    location: Location,
) -> Inspected {
    let access = match regular.creator_only_commitment() {
        Ok(commitment) if commitment == selected.header.security_commitment => {
            AccessPolicy::CreatorOnly
        }
        _ => AccessPolicy::ChangedOrUnproven,
    };
    Inspected {
        name,
        file: regular.file,
        mapping,
        identity: regular.identity,
        header: selected.header,
        location,
        access,
    }
}

fn lock_operation(regular: &Regular, cancellation: &CancellationToken) -> Result<(), Problem> {
    lock_operation_file(&regular.file, cancellation)
}

fn lock_operation_file(file: &File, cancellation: &CancellationToken) -> Result<(), Problem> {
    live_lock::lock_cancellable(file, OPERATION_LOCK, Mode::Exclusive, cancellation)
        .map_err(|error| Problem::sdk(&error))?;
    cancellation.check().map_err(|error| Problem::sdk(&error))
}

pub(super) fn require_bound(
    destination: &Destination,
    header: Header,
    identity: Identity,
    filename_attempt: Option<[u8; 16]>,
) -> Result<(), Problem> {
    if header.reservation_identity != identity.encode() {
        return Err(conflict(
            "reservation self identity does not match its inode",
        ));
    }
    if filename_attempt.is_some_and(|attempt| attempt != header.attempt_id) {
        return Err(conflict("private reservation name has another attempt id"));
    }
    let basename_len =
        u32::try_from(destination.main().bytes().len()).map_err(|_| name_mismatch())?;
    if header.basename_len != basename_len
        || header.basename_commitment != destination.basename_commitment()
    {
        return Err(name_mismatch());
    }
    Ok(())
}

fn map_reservation(file: &File) -> Result<Mapping, ReadError> {
    if file
        .metadata()
        .map_err(error::Error::from)
        .map_err(ReadError::Sdk)?
        .len()
        != FILE_SIZE
    {
        return Err(ReadError::Invalid);
    }
    Mapping::read_write_view(file, FILE_SIZE).map_err(ReadError::Sdk)
}

fn read_selected(mapping: &Mapping) -> Result<Selected, ReadError> {
    reservation::select(
        mapping
            .bytes(0, FILE_SIZE as usize)
            .map_err(ReadError::Sdk)?,
    )
    .map_err(|_| ReadError::Invalid)
}

fn parse_private_name(bytes: &[u8]) -> Option<[u8; 16]> {
    let encoded = bytes.strip_prefix(PREFIX)?.strip_suffix(SUFFIX)?;
    if encoded.len() != ENCODED_ATTEMPT_LEN {
        return None;
    }
    let mut attempt = [0; 16];
    for (slot, pair) in attempt.iter_mut().zip(encoded.chunks_exact(2)) {
        *slot = decode_hex(pair[0])?.checked_mul(16)? + decode_hex(pair[1])?;
    }
    (attempt != [0; 16]).then_some(attempt)
}

fn decode_hex(byte: u8) -> Option<u8> {
    match byte {
        b'0'..=b'9' => Some(byte - b'0'),
        b'a'..=b'f' => Some(byte - b'a' + 10),
        _ => None,
    }
}

fn invalid_private_entry(error: &NamespaceError) -> bool {
    match error {
        NamespaceError::NotRegular
        | NamespaceError::LinkCount(_)
        | NamespaceError::CrossFilesystem => true,
        NamespaceError::IoAt { source, .. } => source.raw_os_error() == Some(libc::ELOOP),
        _ => false,
    }
}

fn strict_record(error: ReadError) -> Problem {
    match error {
        ReadError::Invalid => Problem::new(
            ErrorCode::Unresolvable,
            None,
            "publication reservation record is not selectable",
        ),
        ReadError::Sdk(error) => Problem::sdk(&error),
    }
}

const fn conflict(detail: &'static str) -> Problem {
    Problem::new(ErrorCode::Conflict, None, detail)
}

const fn name_mismatch() -> Problem {
    Problem::new(
        ErrorCode::DestinationNameMismatch,
        None,
        "reservation belongs to another destination name",
    )
}

enum ReadError {
    Invalid,
    Sdk(error::Error),
}

File: v4/go/internal/publication/reservation_inspection.go
// Exact discovery of one restart-authoritative publication reservation
// (Rust publication/reservation_inspection.rs, 451 lines read in
// full). The machine finds the canonical coordination twin or scans
// the private reservation names, proves the bound evidence, takes the
// operation lock, and returns the inspected owner; the resolver and
// cleanup slices compose it. All record reads run over mapped views.
// The Rust gc_barrier availability calls are #[cfg(windows)] and
// compile to nothing on POSIX (Phase-2 GC surface), so they are
// absent here like every earlier slice.

package publication

import (
	"errors"
	"os"

	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/live"
	"github.com/firehol/iprange/v4/go/internal/mapping"
	"github.com/firehol/iprange/v4/go/internal/security"
)

// errReservationInvalid is the non-selectable reservation evidence
// marker (Rust ReadError::Invalid; strict_record maps it to the
// Unresolvable class).
var errReservationInvalid = &format.Error{Code: format.CodeUnresolvable, Detail: "publication reservation record is not selectable"}

// inspectedReservation is one discovered publication reservation with
// its lock held (Rust Inspected).
type inspectedReservation struct {
	name     string
	file     *os.File
	mapping  *mapping.Mapping
	identity live.FileIdentity
	header   reservationHeader
	location reservationLocation
	access   AccessPolicy
}

// Close releases the inspected reservation resources (Rust drop; the
// resolver and cleanup machines consume the same owner fields).
func (r *inspectedReservation) Close() error {
	return closeReservationOwner(r.file, r.mapping)
}

// verify re-proves the inspected reservation at its inspected
// position (Rust Inspected::verify).
func (r *inspectedReservation) verify(destination *destination) error {
	if err := destination.directory().Verify(); err != nil {
		return err
	}
	name := r.name
	if r.location == reservationLocationCanonical {
		name = destination.coordinationName()
	}
	if err := destination.directory().VerifyName(name, r.identity); err != nil {
		return err
	}
	selected, err := readSelected(r.mapping)
	if err != nil {
		return strictRecord(err)
	}
	if selected.header != r.header {
		return conflictProblem("publication reservation changed after inspection")
	}
	return nil
}

// unlockOperation releases the operation lock of one inspected
// reservation (Rust Inspected::unlock_operation).
func (r *inspectedReservation) unlockOperation() error {
	return live.UnlockFile(r.file, reservationOperationLock)
}

// relockOperation re-acquires the operation lock and re-proves the
// inspected reservation (Rust Inspected::relock_operation).
func (r *inspectedReservation) relockOperation(destination *destination, check func() error) error {
	if err := lockOperationFile(r.file, check); err != nil {
		return err
	}
	return r.verify(destination)
}

// discoverReservation finds the restart-authoritative reservation of
// one destination: the canonical coordination twin first, then one
// bound private reservation (Rust discover).
func discoverReservation(destination *destination, check func() error) (*inspectedReservation, error) {
	inspected, err := inspectedCanonical(destination, check)
	if err != nil {
		return nil, err
	}
	if inspected != nil {
		return inspected, nil
	}
	found, err := scanPrivateReservations(destination, check)
	if err != nil {
		return nil, err
	}
	if err := destination.directory().RequireAbsent(destination.coordinationName()); err != nil {
		if found != nil {
			_ = found.Close()
		}
		return nil, conflictProblem("coordination changed during reservation scan")
	}
	return found, nil
}

// inspectedCanonical inspects the coordination twin if it exists
// (Rust canonical).
func inspectedCanonical(destination *destination, check func() error) (*inspectedReservation, error) {
	if err := live.Checkpoint(check); err != nil {
		return nil, err
	}
	regular, err := openCanonicalRegular(destination)
	if err != nil {
		return nil, err
	}
	if regular == nil {
		return nil, nil
	}
	return inspectCanonicalReservation(destination, regular, check)
}

// exactPrivateReservation inspects the private reservation of one
// caller-supplied header (Rust exact_private).
func exactPrivateReservation(destination *destination, expected reservationHeader, check func() error) (inspected *inspectedReservation, err error) {
	if err := live.Checkpoint(check); err != nil {
		return nil, err
	}
	name, err := destination.reservationName(expected.attemptID)
	if err != nil {
		return nil, err
	}
	regular, err := destination.directory().OpenRegular(name, true)
	if err != nil {
		return nil, err
	}
	if regular == nil {
		return nil, nil
	}
	// ownership: the mapped view and the descriptor stay owned by this
	// function until the successful return transfers them to the
	// inspected owner; every error and skip path closes both, exactly
	// like the Rust drop of mapping and regular.
	var mapped *mapping.Mapping
	defer func() {
		if inspected == nil {
			_ = closeReservationOwner(regular.File, mapped)
		}
	}()
	if err := lockOperation(regular, check); err != nil {
		return nil, err
	}
	if err := destination.directory().VerifyName(name, regular.Identity); err != nil {
		return nil, err
	}
	mapped, err = mapReservation(regular.File)
	if err != nil {
		return nil, strictRecord(err)
	}
	selected, err := readSelected(mapped)
	if err != nil {
		return nil, strictRecord(err)
	}
	if err := requireBound(destination, selected.header, regular.Identity, &expected.attemptID); err != nil {
		return nil, err
	}
	if selected.header != expected {
		return nil, conflictProblem("caller result and private reservation disagree")
	}
	return inspectedReservationOf(name, regular, mapped, selected, reservationLocationPrivate), nil
}

// inspectCanonicalReservation proves one canonical coordination twin
// (Rust inspect_canonical).
func inspectCanonicalReservation(destination *destination, regular *live.RegularFile, check func() error) (inspected *inspectedReservation, err error) {
	// ownership: the regular descriptor and the mapped view stay owned
	// by this function until the successful return transfers them to
	// the inspected owner; every error path closes both, exactly like
	// the Rust drop of regular and mapping.
	var mapped *mapping.Mapping
	defer func() {
		if inspected == nil {
			_ = closeReservationOwner(regular.File, mapped)
		}
	}()
	mapped, err = mapReservation(regular.File)
	if err != nil {
		return nil, strictRecord(err)
	}
	selected, err := readSelected(mapped)
	if err != nil {
		return nil, strictRecord(err)
	}
	if err := lockOperation(regular, check); err != nil {
		return nil, err
	}
	rechecked, err := readSelected(mapped)
	if err != nil {
		return nil, strictRecord(err)
	}
	if rechecked != selected {
		return nil, conflictProblem("publication reservation changed while acquiring its lock")
	}
	if err := requireBound(destination, selected.header, regular.Identity, nil); err != nil {
		return nil, err
	}
	privateName, err := destination.reservationName(selected.header.attemptID)
	if err != nil {
		return nil, err
	}
	if err := finishCanonicalTransition(destination, privateName, regular.Identity); err != nil {
		return nil, err
	}
	if err := destination.directory().VerifyName(destination.coordinationName(), regular.Identity); err != nil {
		return nil, err
	}
	rechecked, err = readSelected(mapped)
	if err != nil {
		return nil, strictRecord(err)
	}
	if rechecked != selected {
		return nil, conflictProblem("publication reservation changed during inspection")
	}
	if err := destination.directory().Verify(); err != nil {
		return nil, err
	}
	return inspectedReservationOf(privateName, regular, mapped, selected, reservationLocationCanonical), nil
}

// scanPrivateReservations scans the retained directory for one bound
// private reservation (Rust scan_private).
func scanPrivateReservations(destination *destination, check func() error) (*inspectedReservation, error) {
	var found *inspectedReservation
	err := destination.directory().Scan(func(bytes []byte) error {
		if err := live.Checkpoint(check); err != nil {
			return err
		}
		attemptID, ok := privateAttempt(reservationPrefix, bytes)
		if !ok {
			return nil
		}
		name := string(bytes)
		// Rust Name::new validates the component shape; a name that
		// parsed as the exact private reservation shape is never
		// empty, never "." or "..", and carries no separator or NUL.
		candidate, err := inspectPrivateReservation(destination, name, attemptID, check)
		if err != nil {
			return err
		}
		if candidate == nil {
			return nil
		}
		if found != nil {
			// The second candidate is owned and must close before the
			// conflict return, exactly like Rust drops it.
			_ = candidate.Close()
			return conflictProblem("multiple bound private publication reservations exist")
		}
		found = candidate
		return nil
	})
	if err != nil {
		// The first candidate is owned until this point; the conflict
		// arm above already closed the second one (Rust drops both on
		// the error return).
		if found != nil {
			_ = found.Close()
		}
		return nil, err
	}
	return found, nil
}

// inspectPrivateReservation proves one private reservation candidate
// (Rust inspect_private). Entries that cannot be a bound reservation
// are skipped, exactly like Rust.
func inspectPrivateReservation(destination *destination, name string, attemptID [16]byte, check func() error) (inspected *inspectedReservation, err error) {
	regular, err := destination.directory().OpenRegular(name, true)
	switch {
	case err != nil && invalidPrivateEntry(err):
		return nil, nil
	case err != nil:
		return nil, err
	case regular == nil:
		return nil, nil
	}
	// ownership: same rule as inspectCanonicalReservation - the skip
	// returns are Ok(None) and must close the opened regular and the
	// mapped view exactly like Rust drops them on the None arm.
	var mapped *mapping.Mapping
	defer func() {
		if inspected == nil {
			_ = closeReservationOwner(regular.File, mapped)
		}
	}()
	mapped, err = mapReservation(regular.File)
	if err != nil {
		if errors.Is(err, errReservationInvalid) {
			return nil, nil
		}
		return nil, strictRecord(err)
	}
	selected, err := readSelected(mapped)
	if err != nil {
		if errors.Is(err, errReservationInvalid) {
			return nil, nil
		}
		return nil, strictRecord(err)
	}
	if err := requireBound(destination, selected.header, regular.Identity, &attemptID); err != nil {
		return nil, nil
	}
	if err := lockOperation(regular, check); err != nil {
		return nil, err
	}
	if err := destination.directory().VerifyName(name, regular.Identity); err != nil {
		return nil, conflictProblem("private reservation changed during inspection")
	}
	rechecked, err := readSelected(mapped)
	if err != nil {
		return nil, strictRecord(err)
	}
	if rechecked != selected {
		return nil, conflictProblem("private reservation changed during inspection")
	}
	return inspectedReservationOf(name, regular, mapped, selected, reservationLocationPrivate), nil
}

// requireBound proves the reservation binds this destination, inode,
// and optional filename attempt (Rust require_bound).
func requireBound(destination *destination, header reservationHeader, identity live.FileIdentity, filenameAttempt *[16]byte) error {
	if header.reservationIdentity != reservationIdentityBytes(identity) {
		return conflictProblem("reservation self identity does not match its inode")
	}
	if filenameAttempt != nil && *filenameAttempt != header.attemptID {
		return conflictProblem("private reservation name has another attempt id")
	}
	// The basename length is bounded by the destination name-max proof
	// at bind, so the Rust try_from overflow arm is unreachable in Go.
	basenameLen := uint32(len(destination.mainName()))
	if header.basenameLen != basenameLen || header.basenameCommitment != destination.basenameCommitmentValue() {
		return destinationNameMismatchProblem()
	}
	return nil
}

// mapReservation maps one reservation file after the exact-size proof
// (Rust map_reservation; a wrong file size is the Invalid marker).
func mapReservation(file *os.File) (*mapping.Mapping, error) {
	st, err := file.Stat()
	if err != nil {
		return nil, &format.Error{Code: format.CodeIO, Detail: "stat: " + err.Error()}
	}
	if uint64(st.Size()) != reservationFileSize {
		return nil, errReservationInvalid
	}
	return mapping.MapFile(file, reservationFileSize, true)
}

// readSelected decodes the selectable reservation record of one
// mapped view (Rust read_selected; a select refusal is the Invalid
// marker).
func readSelected(m *mapping.Mapping) (selectedReservation, error) {
	bytes, err := m.View(0, reservationFileSize)
	if err != nil {
		return selectedReservation{}, err
	}
	selected, err := selectReservation(bytes)
	if err != nil {
		return selectedReservation{}, errReservationInvalid
	}
	return selected, nil
}

// inspectedReservationOf builds the inspected owner with the
// creator-only evidence classification (Rust inspected).
func inspectedReservationOf(name string, regular *live.RegularFile, m *mapping.Mapping, selected selectedReservation, location reservationLocation) *inspectedReservation {
	access := AccessPolicyChangedOrUnproven
	if commitment, err := security.CreatorOnlyCommitment(regular.File); err == nil && commitment == selected.header.securityCommitment {
		access = AccessPolicyCreatorOnly
	}
	return &inspectedReservation{
		name:     name,
		file:     regular.File,
		mapping:  m,
		identity: regular.Identity,
		header:   selected.header,
		location: location,
		access:   access,
	}
}

// lockOperation takes the exclusive operation lock of one opened
// artifact (Rust lock_operation).
func lockOperation(regular *live.RegularFile, check func() error) error {
	return lockOperationFile(regular.File, check)
}

// lockOperationFile takes the exclusive operation lock with
// cancellation and re-checks the caller cancellation (Rust
// lock_operation_file).
func lockOperationFile(file *os.File, check func() error) error {
	if err := live.LockFileCancellable(file, reservationOperationLock, live.LockExclusive, check); err != nil {
		return err
	}
	return live.Checkpoint(check)
}

// invalidPrivateEntry classifies one entry that cannot be a private
// reservation (Rust invalid_private_entry).
func invalidPrivateEntry(err error) bool {
	nerr, ok := live.AsNamespaceError(err)
	if !ok {
		return false
	}
	switch nerr.Kind {
	case live.NamespaceNotRegular, live.NamespaceLinkCount, live.NamespaceCrossFilesystem:
		return true
	case live.NamespaceIoAt:
		return live.IsNofollowSymlink(nerr.Err)
	}
	return false
}

// strictRecord maps one reservation read refusal (Rust strict_record:
// Invalid to the fixed Unresolvable problem, SDK errors preserve
// their class with the fixed SDK detail).
func strictRecord(err error) *format.Error {
	if errors.Is(err, errReservationInvalid) {
		return errReservationInvalid
	}
	return sdkProblem(err)
}

// conflictProblem builds the fixed conflict class of one inspection
// refusal (Rust conflict).
func conflictProblem(detail string) error {
	return problem(format.CodeConflict, detail)
}

// destinationNameMismatchProblem builds the fixed destination-name
// mismatch class (Rust name_mismatch).
func destinationNameMismatchProblem() error {
	return problem(format.CodeDestinationNameMismatch, "reservation belongs to another destination name")
}
